pkg/aggregator/time_sampler.go

```diff
@@ -75,11 +134,121 @@ func (s *TimeSampler) addSample(metricSample *metrics.MetricSample, timestamp fl
 
 		// Add sample to bucket
 		if err := bucketMetrics.AddSample(contextKey, metricSample, timestamp, s.interval, nil); err != nil {
-			log.Debugf("Ignoring sample '%s' on host '%s' and tags '%s': %s", metricSample.Name, metricSample.Host, metricSample.Tags, err)
+			log.Debugf("TimeSampler #%d Ignoring sample '%s' on host '%s' and tags '%s': %s", s.id, metricSample.Name, metricSample.Host, metricSample.Tags, err)
 		}
 	}
 }
 
+// Stop stops the time sampler. It can't be re-used after being stop,
+// use NewTimeSampler instead.
+func (s *TimeSampler) Stop() {
+	s.stopChan <- struct{}{}
+}
+
+// We process all receivend samples in the `select`, but we also process a flush action,
+// meaning that the time sampler will not process any sample while it is flushing.
+// Note that it was the same design in the BufferedAggregator (but at the aggregator level,
+// not sampler level).
+// If we want to move to a design where we can flush while we are processing samples,
+// we could consider implementing double-buffering or locking for every sample reception.
+func (s *TimeSampler) processLoop(flushInterval time.Duration) {
+	ticker := time.NewTicker(flushInterval)
+	for {
+		select {
+		case <-s.stopChan:
+			return
+		case ms := <-s.samples:
+			// do this telemetry here and not in the samplers goroutines, this way
```

Comment:
But this is in the sampler goroutine though?

Reply:
Yup and effectively not much I can do about this actually, however, they did not show up in benchmarks so we are fine, I'm just removing the comment.